[[ Script ]] Several libscript API additions

runrevmark opened the pull request:
This patch adds:

```
- MCScriptCopyConstantsOfModule
- MCScriptQueryConstantOfModule
- MCScriptCopyHandlerOfInstance
- MCScriptCopyHandlerParametersOfModule
```

This patch renames:

```
- MCScriptQueryHandlerOfModule to MCScriptQueryHandlerSignatureOfModule
```

runrevmark committed:
[[ Script ]] Add accessors for constants and values

This patch adds CopyConstantsOfModule and QueryConstantOfModule to
the libscript API. These enable fetching the list of public constants
defined in the module, and the value of a public constant defined
in a module.

runrevmark committed:
[[ Script ]] Add ability to query a handler value bound to an instance

This patch adds a QueryHandlerOfInstance API which allows creation
of a handler-ref which will invoke a specified handler within a given
instance.

runrevmark committed:
[[ Script ]] Add the ability to query the parameter names of a handler

This patch adds CopyHandlerParametersOfModule API which allows
fetching the list of parameter names of the specified handler (as long
as debug information hasn't been stripped from the module).

It also renames CopyHandlerOfModule to CopyHandlerSignatureOfModule
and updates uses of the function.

runrevmark committed:
[[ Script ]] Tweak naming of some libscript functions

This patch renames some of the libscript functions to have more
appropriate names.

Specifically:

. MCScriptListDependencyNamesOfModule
. MCScriptListConstantNamesOfModule
. MCScriptListPropertyNamesOfModule
. MCScriptListEventNamesOfModule
. MCScriptListHandlerNamesOfModule
. MCScriptListHandlerParameterNamesOfModule
. MCScriptGetPropertyInInstance
. MCScriptSetPropertyInInstance
. MCScriptCallHandlerInInstance
. MCScriptCallHandlerInInstanceIfFound
. MCScriptEvaluateHandlerBindingInInstance

peter-b commented on libscript/src/script-module.cpp:
```diff
@@ -983,78 +1014,115 @@ void MCScriptReleaseRawModule(MCScriptModule *p_module)
+__MCScriptLookupDefinitionInModule(MCScriptModuleRef self,
+								   MCScriptDefinitionKind p_kind,
+								   MCNameRef p_name,
+								   MCScriptDefinition*& r_definition)
```
In order to avoid an awkward cast in all the callers, I wonder if this could be written as:

``` C++
template <typename T, MCScriptDefinitionKind K>
__MCScriptLookupDefinitionInModule(MCScriptModuleRef self, MCNameRef p_name, T*& r_definition)
{
    /* ... */
}
```

peter-b replied:
To simplify the callers further, could you move the following call that appears in all callers into this function?

``` C++
 __MCScriptValidateObjectAndKind__(self, kMCScriptObjectKindModule);
```

peter-b commented on libscript/src/script-module.cpp:
```diff
@@ -995,6 +995,42 @@ bool MCScriptQueryHandlerOfModule(MCScriptModuleRef self, MCNameRef p_handler, /
+	
+    if (!MCScriptLookupHandlerDefinitionInModule(self, p_handler, t_def))
+	{
+        return false; // TODO - throw error
```
Is there any reason why these errors can't be implemented now?

peter-b reviewed (commented):
Looks good to me, with a minor query about whether the implementation could be deduplicated a bit.

runrevmark committed:
[[ Script ]] Cleanup

This commit implements the missing error throwing in the Query
functions and uses a template to implement the LookupDefinition
functions.

peter-b commented:
@livecode-vulcan review ok 40529a2

livecode-vulcan commented:
💙 review by @peter-b ok 40529a2f1af46fdc7b803f5d512857160788b6ec

livecode-vulcan commented:
😎 test success 40529a2f1af46fdc7b803f5d512857160788b6ec
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/217)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/221)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/217)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/217)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/221)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/217)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/217)

runrevmark committed:
[[ Script ]] Correct functionality of Query methods in MCScript

The definition querying methods should return false if the module is
not usable or the definition could not be found - they should not
throw (as they are used to check whether a module includes a given
named definition!).

peter-b commented:
@livecode-vulcan review ok c4f30ab

livecode-vulcan commented:
💙 review by @peter-b ok c4f30abceb4e552533692f0c36df901bf8f23ac4

livecode-vulcan commented:
😎 test success c4f30abceb4e552533692f0c36df901bf8f23ac4
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/219)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/223)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/219)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/219)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/223)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/219)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/219)
